tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,143 @@
+
+from .helpers import access_all_dashboards, get_relation_data
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+# FIXME: update this to 'stable' when `pre-upgrade-check` is released to 'stable'
+CHANNEL = "edge"
+
+OPENSEARCH_APP_NAME = "opensearch"
+OPENSEARCH_CONFIG = {
+    "logging-config": "<root>=INFO;unit=DEBUG",
+    "cloudinit-userdata": """postruncmd:
+        - [ 'sysctl', '-w', 'vm.max_map_count=262144' ]
+        - [ 'sysctl', '-w', 'fs.file-max=1048576' ]
+        - [ 'sysctl', '-w', 'vm.swappiness=0' ]
+        - [ 'sysctl', '-w', 'net.ipv4.tcp_retries2=5' ]
+    """,
+}
+TLS_CERTIFICATES_APP_NAME = "self-signed-certificates"
+
+NUM_UNITS_APP = 3
+NUM_UNITS_DB = 2
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.charm
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Deploying all charms required for the tests, and wait for their complete setup to be done."""
+
+    pytest.charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(pytest.charm, application_name=APP_NAME, num_units=NUM_UNITS_APP)
+    await ops_test.model.set_config(OPENSEARCH_CONFIG)
+    # Pinning down opensearch revision to the last 2.10 one
+    # NOTE: can't access 2/stable from the tests, only 'edge' available
+    await ops_test.model.deploy(OPENSEARCH_APP_NAME, channel="2/edge", num_units=NUM_UNITS_DB)
+
+    config = {"ca-common-name": "CN_CA"}
+    await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="stable", config=config)
+
+    await ops_test.model.wait_for_idle(
+        apps=[TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    # Relate it to OpenSearch to set up TLS.
+    await ops_test.model.relate(OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == NUM_UNITS_APP
+        )
+        await ops_test.model.wait_for_idle(apps=[APP_NAME], timeout=1000, idle_period=30)
+
+    assert ops_test.model.applications[APP_NAME].status == "blocked"
+
+    pytest.relation = await ops_test.model.relate(OPENSEARCH_APP_NAME, APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, APP_NAME], status="active", timeout=1000
+    )
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_in_place_upgrade_http(ops_test: OpsTest):
+    leader_unit = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        if await unit.is_leader_from_status():
+            leader_unit = unit
+    assert leader_unit
+
+    action = await leader_unit.run_action("pre-upgrade-check")
+    await action.wait()
+
+    # ensure action completes
+    time.sleep(10)
```

Comment:
this should not be needed, `action.wait` is a blocking operation

Reply:
Done.